stan/math/prim/scal/meta/is_detected.hpp

```diff
@@ -0,0 +1,23 @@
+#ifndef STAN_MATH_PRIM_SCAL_META_IS_DETECTED_HPP
+#define STAN_MATH_PRIM_SCAL_META_IS_DETECTED_HPP
+
+#include <stan/math/prim/scal/meta/void_t.hpp>
+#include <type_traits>
+
+namespace stan {
+
+template <typename, template <typename> class, typename = void>
+struct is_detected : std::false_type {};
+
+/**
+ * Checks whether a valid type is detected. Most commonly used to detect
+ *  attributes of objects.
+ * @tparam T The type to be checked.
+ * @tparam Op a template template type which attempts to define a member
+ *  of an object.
```

Comment:
Is this a typo ('template template')? Or is it actually a template template type?

Reply:
It's actually a template template type. So this wouldn't work for non-templated detectors. This template param is for the detector for the op so it let's us write

```cpp
template <typename T>
using operator_bracket_t = decltype(std::declval<T>()[int{}]);
```

```cpp
template <typename T>
struct is_vector_like
    : bool_constant<is_detected<T, internal::operator_bracket_t>::value> {};
```

Where `operator_bracket_t` is not fully a fully constructed type

I'm pretty sure it needs to be like this so that the type construction is delayed until `void_t<Op<T>>` in 

```cpp
template <typename T, template <typename...> class Op>
struct is_detected<T, Op, void_t<Op<T>>> : std::true_type {};
```